### ISSUE TITLE
Automatic constructor dependencies

### ISSUE THREAD

**Issue by SamMousa** [yiisoft/di#51]
Currently DI supports specifically setting constructor arguments via the following syntax:
```php
return [
    'full_definition' => [
        '__class' => EngineMarkOne::class,
        '__construct()' => [42], 
        'argName' => 'value',
        'setX()' => [42],
    ],
];
```

Consider the following scenario. I have a cache component.
```php
class CoolCache {
    public function __construct(CacheInterface $handler) { ... }
}

return [
    'my_cool_cache' => [
        '__class' => CoolCache::class
    ]
];
```

The DI container will require us to define a concrete class for the `CacheInterface`, otherwise it cannot instantiate my cool cache component. Currently we could specify this as a constructor argument, but this is very cumbersome.
Alternatively we specify it globally:
```php

return [
    'my_cool_cache' => [
        '__class' => CoolCache::class
    ],
    CacheInterface::class => SomeCacheHandler::class
];

```
This works great, the DI container will figure it out.

Now consider a more advanced situation where I have 2 caches, and I want them to use different handlers:

```php
return [
    'my_cool_cache' => [
        '__class' => CoolCache::class
    ],
    'another_cool_cache' => [
        '__class' => CoolCache::class
    ],
    CacheInterface::class => SomeCacheHandler::class
];
```
We have a problem, we either cannot specify what we want, or we need to specify it too explicitly.

My proposal is add support for contextual configuration of the DI container:
```php
return [
    'my_cool_cache' => [
        '__class' => CoolCache::class
    ],

    'another_cool_cache' => [
        '__class' => CoolCache::class

        '__overrides' => [
            CacheInterface::class => AnotherCacheHandler::class

        ]
    ],
    CacheInterface::class => SomeCacheHandler::class
];
```

The idea is simple yet powerful. 
Basically when resolving `my_cool_cache`, we first load all overrides.
Then we proceed as normal.
Then we revert the overrides.

`__overrides` has the exact same syntax as the configuration array of the DI container.

Benefits:
    - No need to specify argument names explicitly, we're just specifying how dependencies should be resolved IF they are needed.
    - The implementing class can properly use constructor injection.
    - The implementing class no longer needs to know about DI at all.

**Comment by hiqsol** [comment]
Looks good to me.
I like the idea to have different interface implementations for particular cases.

**Comment by SamMousa** [comment]
What about an implementation that uses `clone`?

**Comment by samdark** [comment]
@SamMousa what do you mean?

**Comment by SamMousa** [comment]
Clone the container, override definitions, instantiate object.

This won't work k for composite containers though, I think

**Comment by samdark** [comment]
Don't know. I don't see how that helps in implementing it.
